File: LeetCode/[M] - 437. Path Sum III/solution.py
from typing import Optional


# TreeNode - Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        freq = {0: 1}
        self.cnt = 0

        def searchSum(node, ps):
            if not node:
                return
            cs = ps + node.val
            x = cs - targetSum
            if x in freq:
                self.cnt += freq[x]
            if cs in freq:
                freq[cs] += 1
            else:
                freq[cs] = 1
            searchSum(node.left, cs)
            searchSum(node.right, cs)
            freq[cs] -= 1

        searchSum(root, 0)
        return self.cnt

File: LeetCode/[M] - 437. Path Sum III/solution.go
package main

// TreeNode - Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) int {
	freq := map[int]int{0: 1}
	res := 0

	var dfs func(node *TreeNode, ps int)
	dfs = func(node *TreeNode, ps int) {
		if node == nil {
			return
		}
		cs := ps + node.Val
		x := cs - targetSum

		res += freq[x]
		freq[cs] += 1

		dfs(node.Left, cs)
		dfs(node.Right, cs)

		freq[cs] -= 1
	}
	dfs(root, 0)
	return res
}
